### DISCUSSION TITLE
How to optimize the direction of directional light

### DISCUSSION THREAD

**Discussion by jensenzhoujh** [mitsuba-renderer/mitsuba3#660] [Q&A]
Hi @,

I want to optimize the direction of directional light but find that the `direction` and `to_world` params for `directional` light are not differentiable. I've tried using `area directional` light with a fairly large ceiling and twisted the direction of light by changing the normals of ceiling vertices. However, the object can not be properly illuminated even with a large radiance value. I am wondering whether there are recommended workarounds that will serve my purposes.

Thanks

**Comment by njroussel** [comment]
Hi @shallowtoil 

This is unfortunate, but I don't believe you can optimize the vertex normals to achieve a similiar result to optimizing a `directional` emitter (if that were possible).
We don't have any integrators that can handle differentiation of delta/dirac emittters or BSDFs for that matter.

Optimizing the position of your emitter is possible in most cases. I would recommend trying to approximate your scene/setup with an emitter which you'd be moving.

**Comment by jensenzhoujh** [comment]
Hi @njroussel,

Thanks for the response. By optimizing the position of your emitter, do you mean to adjust the orientation of my ceiling? Or use a point light? In my case, a distant directional source is the best since I want everything in my scene (suppose it's a fairly large scene) to be uniformly illuminated. Another workaround I can think of is to instead of adjusting the direction of the light, optimize the positions of objects and the sensor since all those are differentiable.
